### PR TITLE
Rework type support implementation and some of the generated marshal helper methods

### PR DESCRIPTION
**Pull request by jwillemsen**
This PR reworks two areas

1. The type support implementation is reworked to use a C++ class instead of a C++ template. The big advantage of this is that the implementation of the type support, type specific data reader, type specific data writer and a lot of dependent templates is now not seen anymore in the user code.
2. Introduced a new MarshalTraits<> which we specialize for each type. Using this new trait it is possible to retrieve whether the type and its key are bounded or not without the need to create an instance of the type. This saves some unnecessary temporary objects but did lead to some more work in unit tests which check the validated code.

In a future step we could extend MarshalTraits<> with gen_max_marshaled_size for the type and the key which could save us some more temporary instances.

### CODE REVIEW

**Comment by mitza-oci** [comment]
In case you didn't see it on Travis:
```PubDriver.cpp:893:22: error: ‘<::’ cannot begin a template-argument list```

**Comment by jwillemsen** [comment]
Fixed

File: dds/idl/marshal_generator.cpp [hunk 1]
```diff
@@ -1315,6 +1303,13 @@ bool marshal_generator::gen_struct(AST_Structure*, UTL_ScopedName* name,
       if (first) be_global->impl_ << intro << "  return true;\n";
       else be_global->impl_ << intro << "  return " << expr << ";\n";
     }
+
+    be_global->header_ <<
+      "template <>\n"
+      "struct MarshalTraits<" << cxx << "> {\n"
+      "inline static bool gen_is_bounded_size() { return " << (is_bounded_struct ? "true" : "false") << "; }\n"
```
Comment (mitza-oci) [inline dds/idl/marshal_generator.cpp]: "inline" isn't needed here, all member functions defined within the class definition are inline.  Please update in DDSTraits as well.

File: dds/idl/ts_generator.cpp [hunk 2]
```diff
@@ -113,36 +118,116 @@ bool ts_generator::gen_struct(AST_Structure*, UTL_ScopedName* name,
...
+      "  virtual ::DDS::DataReader_ptr create_datareader();\n"
+      "#ifndef OPENDDS_NO_MULTI_TOPIC\n"
+      "  virtual ::DDS::DataReader_ptr create_multitopic_datareader();\n"
+      "#endif /* !OPENDDS_NO_MULTI_TOPIC */\n"
+      "#ifndef OPENDDS_NO_CONTENT_SUBSCRIPTION_PROFILE\n"
+      "  virtual const OpenDDS::DCPS::MetaStruct& getMetaStructForType();\n"
+      "#endif /* !OPENDDS_NO_CONTENT_SUBSCRIPTION_PROFILE */\n"
+      "  virtual bool has_dcps_key();\n"
+      "  const char* default_type_name() const;\n"
+      "  " << short_name << "TypeSupport::_ptr_type _narrow(CORBA::Object_ptr obj);\n"
```
Comment (mitza-oci) [inline dds/idl/ts_generator.cpp]: _narrow() was a static member function in the previous version

File: dds/idl/ts_generator.cpp [hunk 3]
```diff
@@ -113,36 +118,116 @@ bool ts_generator::gen_struct(AST_Structure*, UTL_ScopedName* name,
...
+      "\n"
+      "  virtual ::DDS::DataWriter_ptr create_datawriter();\n"
+      "  virtual ::DDS::DataReader_ptr create_datareader();\n"
+      "#ifndef OPENDDS_NO_MULTI_TOPIC\n"
+      "  virtual ::DDS::DataReader_ptr create_multitopic_datareader();\n"
+      "#endif /* !OPENDDS_NO_MULTI_TOPIC */\n"
+      "#ifndef OPENDDS_NO_CONTENT_SUBSCRIPTION_PROFILE\n"
+      "  virtual const OpenDDS::DCPS::MetaStruct& getMetaStructForType();\n"
+      "#endif /* !OPENDDS_NO_CONTENT_SUBSCRIPTION_PROFILE */\n"
+      "  virtual bool has_dcps_key();\n"
```
Comment (mitza-oci) [inline dds/idl/ts_generator.cpp]: has_dcps_key() doesn't appear in the TypeSupportImpl base class.  I wonder if this was left over and isn't needed any more.

Comment (jwillemsen) [inline dds/idl/ts_generator.cpp]: Looks to be in OpenDDS::DCPS::TypeSupport

File: dds/idl/ts_generator.cpp [hunk 4]
```diff
@@ -113,36 +118,116 @@ bool ts_generator::gen_struct(AST_Structure*, UTL_ScopedName* name,
...
+      "public:\n"
+      "  typedef OpenDDS::DCPS::DDSTraits<" << short_name << "> TraitsType;\n"
+      "  typedef " << short_name << "TypeSupport TypeSupportType;\n"
+      "  typedef " << short_name << "TypeSupport::_var_type _var_type;\n"
+      "  typedef " << short_name << "TypeSupport::_ptr_type _ptr_type;\n"
+      "\n"
+      "  " << short_name << "TypeSupportImpl() {}\n"
+      "  virtual ~" << short_name << "TypeSupportImpl() {}\n"
+      "\n"
+      "  virtual ::DDS::DataWriter_ptr create_datawriter();\n"
```
Comment (jwillemsen) [inline dds/idl/ts_generator.cpp]: It is required here, the namespace-conflict test shows an use case

File: dds/idl/marshal_generator.cpp [hunk 5]
```diff
@@ -1315,6 +1303,13 @@ bool marshal_generator::gen_struct(AST_Structure*, UTL_ScopedName* name,
       if (first) be_global->impl_ << intro << "  return true;\n";
       else be_global->impl_ << intro << "  return " << expr << ";\n";
     }
+
+    be_global->header_ <<
+      "template <>\n"
+      "struct MarshalTraits<" << cxx << "> {\n"
```
Comment (jwillemsen) [inline dds/idl/marshal_generator.cpp]: done

**Comment by jwillemsen** [comment]
retest this please

**Comment by mitza-oci** [comment]
I think there are two remaining review issues:
- [ ] make _narrow() static in ts_generator.cpp
- [ ] bool empty in ts_generator.cpp

**Comment by jwillemsen** [comment]
I thought I fixed those, will check again this evening

**Comment by mitza-oci** [comment]
Oh, I think I must have just missed it with the others.  Looks like this is ready to merge.

**Pull request merged**